### ISSUE TITLE
Can we only sample on the feasible region?

### ISSUE THREAD

**Issue by yuanquan010** [pytorch/botorch#2078]
## Issue description
Hello! I am trying to optimize the coefficients of two functions.  About these functions, there are a series of constraints. These two functions will be used to generate parameters passed into the simulation programs, and the simulation program will calculate two properties.  The objective is to maximize these two properties by adjusting the coefficients of the functions.  
Provide a short description.
My question is how can we involve the constraints in the closed BO loop or ask BO only sample in the feasible region?  It looks like the offered parameters constraint form and result constraint form can not describe my constraint conditions.
## Code example
![image](https://github.com/pytorch/botorch/assets/54976645/40ed5d13-77ad-46fa-8e78-eaf229da2519)
![image](https://github.com/pytorch/botorch/assets/54976645/73ef3bc5-d7bc-4c44-a8c4-0a992aafd21a)
![image](https://github.com/pytorch/botorch/assets/54976645/e9796642-ee85-47f9-a9ec-2cb1c2f0625e)
![image](https://github.com/pytorch/botorch/assets/54976645/b9d46730-bf2b-430a-a670-c1183f40158d)

In the above code example, a1,b1,a2,b2 are the coefficients I need to optimize. I tried to add a constraints judgment function between the sampling and the experiment's complete function, but it didn't work well. 

Is there any other strategy and examples to solve my problem?

**Comment by Balandat** [comment]
I'm trying to understand exactly the nature of your constraints. It looks like while the constraints you are imposing on the parameters in `is_acceptable_with_output` are linear, the way they're being used in `is_acceptable` is nonlinear (as you're multiplying the elements of `values` with the other parameter values). So this unfortunately wouldn't be supported as an explicit linear parameter constraint with the current functionality in Ax. 

One potential option would be to consider each of the 7 `f_n` constraint functions `p_n` as a "black box" output, and then define outcome constraints for those (i.e. `p_i <= 16` and `p_i >= 3` for each `i`) via the `outcome_constraints` here https://github.com/facebook/Ax/blob/main/ax/service/ax_client.py#L252. The challenge with this though is that these are a lot of outcome constraints to model, so this will likely be very slow. The requirement that `sum(values) != 60` is something that will also be hard to address efficiently as this is a nonlinear equality constraint.

The bottom line is that unfortunately these are very complex constraints that aren't easy to incorporate into a Bayesian Optimization loop in Ax today. I wonder if it's possible to relax some of these constraints or reformulate them to something that is easier to handle?

Another option would be to drop down one level and use BoTorch instead of Ax, which has (experimental) support for nonlinear constraints on the parameters, see https://github.com/pytorch/botorch/blob/8cdc595b6043e1635f3e5360af2fda729b9ddefd/botorch/optim/optimize.py#L463-L470. That would require manually setting up the optimization loop though.

**Comment by esantorella** [comment]
Closing due to inactivity. Feel free to reopen if you have further questions.
